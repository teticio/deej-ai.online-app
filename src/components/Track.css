.highlight {
  border: 2px solid white;
}
