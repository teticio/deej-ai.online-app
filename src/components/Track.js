import React, { PureComponent, Suspense, useState } from 'react';
import { View, Spinner, IFrame } from './Platform';

function createResource(pending) {
  let error, response;
  pending.then(r => response = r).catch(e => error = e);
  return {
    read() {
      if (error) throw error;
      if (response) return response;
      throw pending;
    }
  };
}

// Avoid unncessary re-renders
class SpotifyTrackWidget extends PureComponent {
  render() {
    const { track_id, resource } = this.props;
    const iframe = (resource) ? resource.read() : null;

    return (
      <IFrame
        title={track_id}
        width='100%'
        height={80}
        srcdoc={Buffer.from(iframe.data, 'base64').toString()}
      />
    );
  }
}

export default function Track({
  track_id
}) {
  if (typeof Track.cache == 'undefined') {
    Track.cache = {};
  }

  const [resource, ] = useState((track_id in Track.cache) ? Track.cache[track_id] :
    createResource(new Promise(resolves => {
      fetch(`${process.env.REACT_APP_API_URL}/widget?track_id=${track_id}`)
        .then(response => (response.status === 200) ? response.text() : '')
        .then(data => resolves({ data: data }))
        .catch(error => console.error('Error:', error));
    })));
  Track.cache[track_id] = resource;

  return (
    <Suspense fallback={
      <View
        style={{
          display: 'flex',
          height: 80,
          width: '100%',
          justifyContent: 'center',
          alignItems: 'center'
        }}
      >
        <Spinner animation='border' />
      </View>
    } >
      <SpotifyTrackWidget
        track_id={track_id}
        resource={resource}
      />
    </Suspense>
  );
}
