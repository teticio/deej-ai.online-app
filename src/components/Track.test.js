import { render, screen, waitFor } from "@testing-library/react";
import Track from "./Track";

jest.setTimeout(20000);

test("Test track widget", async () => {
  render(<Track track_id="3KZMPnlyKTDV4W7oNs7Vbk" testing={true} />);
  const widget = await waitFor(() => screen.getByTitle("3KZMPnlyKTDV4W7oNs7Vbk"), { timeout: 20000 });
  const strippedWidget = widget.srcdoc.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, "");
  console.log(strippedWidget);
  expect(strippedWidget).toMatchSnapshot();
});
