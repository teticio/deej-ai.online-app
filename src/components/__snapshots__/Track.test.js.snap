// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`track widget 1`] = `
"<!DOCTYPE html>

<html dir=\\"ltr\\">
<head>
<meta charset=\\"utf-8\\"/>
<link href=\\"https://open.scdn.co/cdn/images/favicon32.8e66b099.png\\" rel=\\"icon\\" sizes=\\"32x32\\" type=\\"image/png\\"/>
<link href=\\"https://open.scdn.co/cdn/images/favicon16.c498a969.png\\" rel=\\"icon\\" sizes=\\"16x16\\" type=\\"image/png\\"/>
<title>Spotify</title>
<meta content=\\"width=device-width, initial-scale=1.0, user-scalable=1.0, minimum-scale=1.0, maximum-scale=1.0\\" name=\\"viewport\\"/>
<link href=\\"https://open.spotify.com/track/3KZMPnlyKTDV4W7oNs7Vbk\\" rel=\\"canonical\\"/>
<style type=\\"text/css\\">
        @font-face {
            font-family: 'spotify-circular';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Light.afd9ab26.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Light.2a78c017.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Light.89e4be2e.ttf\\") format('truetype');
            font-weight: 200;
            font-style: normal;
        }
        @font-face {
            font-family: 'spotify-circular';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Book.3466e0ec.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Book.ea8d19db.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Book.a357677a.ttf\\") format('truetype');
            font-weight: 400;
            font-style: normal;
        }
        @font-face {
            font-family: 'spotify-circular';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Bold.8d0a45cc.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Bold.10e93738.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIv3T-Bold.7eb7d0f7.ttf\\") format('truetype');
            font-weight: 600;
            font-style: normal;
        }
        @font-face {
            font-family: 'spotify-circular-arabic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Light.dde45e52.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Light.5c7ed8d8.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Light.f0833f3b.otf\\") format('opentype');
            font-weight: 200;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-arabic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Book.aaab109b.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Book.3ad862e2.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Book.f0effcd1.otf\\") format('opentype');
            font-weight: 400;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-arabic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Bold.317e96e5.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Bold.9ddc7126.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIAraOnly-Bold.f97cb921.otf\\") format('opentype');
            font-weight: 600;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-hebrew';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Light.4884da16.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Light.2128eeac.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Light.2e1df62a.otf\\") format('opentype');
            font-weight: 200;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-hebrew';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Book.cb36b6ec.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Book.ad2b13cd.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Book.fae4aece.otf\\") format('opentype');
            font-weight: 400;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-hebrew';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Bold.ef043f44.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Bold.9f406804.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUIHbrOnly-Bold.57b36278.otf\\") format('opentype');
            font-weight: 600;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-cyrillic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Light.55fa9f4a.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Light.034997d0.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Light.b10eb959.otf\\") format('opentype');
            font-weight: 200;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-cyrillic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Book.6a858994.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Book.918c53b8.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Book.15af6ebe.otf\\") format('opentype');
            font-weight: 400;
            font-style: normal;
            font-display: swap;
        }
        @font-face {
            font-family: 'spotify-circular-cyrillic';
            src: url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Bold.30dc9fb2.woff2\\") format('woff2'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Bold.909126fc.woff\\") format('woff'),
                url(\\"https://open.scdn.co/cdn/fonts/CircularSpUICyrOnly-Bold.67641b90.otf\\") format('opentype');
            font-weight: 600;
            font-style: normal;
            font-display: swap;
        }
        body {
            /* IE scrollbar styles */
            scrollbar-base-color: #181818;
            scrollbar-face-color: #404040;
            scrollbar-3dlight-color: #181818;
            scrollbar-highlight-color: #404040;
            scrollbar-track-color: #181818;
            scrollbar-arrow-color: #404040;
            scrollbar-shadow-color: #181818;
            scrollbar-darkshadow-color: #181818;
        }
        /* Google Chrome (Windows) scrollbar styles */
        ::-webkit-scrollbar {
            width: 20px;
            background: transparent;
        }
        ::-webkit-scrollbar-button {
            display: none;
        }
        ::-webkit-scrollbar-track {
            background: transparent;
        }
        ::-webkit-scrollbar-thumb {
            border-radius: 100px;
            background: #aaaaaacc;
            background-clip: content-box;
            border: solid 8px transparent;
        }
    </style>



</head>
<body>




</body>
</html>
"
`;
